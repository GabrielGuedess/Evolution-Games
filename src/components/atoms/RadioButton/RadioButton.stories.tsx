import { Story, Meta } from '@storybook/react/types-6-0';

import { PlatformIcon } from '../PlatformIcon/PlatformIcon';
import { RadioButton, RadioButtonProps } from './RadioButton';

export default {
  title: 'Atoms/RadioButton',
  component: RadioButton,
  args: {
    title: 'playstation',
    icon: <PlatformIcon platform="ps4" size="small" />,
  },
} as Meta<RadioButtonProps>;

export const Default: Story<RadioButtonProps> = args => (
  <RadioButton {...args} />
);
