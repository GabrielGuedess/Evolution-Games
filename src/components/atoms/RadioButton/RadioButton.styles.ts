import { rgba } from 'polished';

import styled, { css, keyframes } from 'styled-components';

const stretchAnimate = keyframes`
  0% {
    transform: scale(1,1)
  }
  28% {
    transform: scale(1.15,.85)
  }
  50% {
    transform: scale(.9,1.1)
  }
  100% {
    transform: scale(1,1)
  }
`;

export const Wrapper = styled.div``;

export const Radio = styled.input`
  ${({ theme }) => css`
    appearance: none;
    position: relative;
    height: 2rem;
    width: 2rem;
    outline: none;
    margin: 0;
    cursor: pointer;
    border: 0.2rem solid ${rgba('#717B8A', 0.5)};
    background: transparent;
    border-radius: 50%;
    display: grid;
    justify-self: end;
    justify-items: center;
    align-items: center;
    overflow: hidden;
    transition: border 0.5s ease;

    ::before,
    ::after {
      content: '';
      display: flex;
      justify-self: center;
      border-radius: 50%;
    }

    ::before {
      position: absolute;
      width: 100%;
      height: 100%;
      background: ${theme.colors.mainBg};
      z-index: 1;
      opacity: var(--opacity, 1);
    }

    ::after {
      position: relative;
      width: calc(100% / 2);
      height: calc(100% / 2);
      background: ${theme.colors.primary};
      top: var(--y, 100%);
      transition: top 0.5s cubic-bezier(0.48, 1.97, 0.5, 0.63);
    }

    :checked {
      border: 0.2rem solid ${theme.colors.primary};
      box-shadow: 0 0.5rem 1.5rem ${rgba(theme.colors.primary, 0.25)};

      ::after {
        --y: 0%;
        animation: ${stretchAnimate} 0.3s ease-out 0.17s;
      }

      ::before {
        --opacity: 0;
      }

      ~ input[type='radio'] {
        ::after {
          --y: -100%;
        }
      }
    }
    :not(:checked) {
      ::before {
        --opacity: 1;
        transition: opacity 0s linear 0.5s;
      }
    }
  `}
`;

export const WrapperInfos = styled.div`
  cursor: pointer;
  display: flex;
  align-items: center;
  gap: 1rem;

  span {
    text-transform: capitalize;
    font-size: 1.4rem;
  }
`;

export const Label = styled.label`
  ${({ theme }) => css`
    cursor: pointer;
    display: flex;
    gap: 2rem;
    width: 100%;
    height: 100%;
    padding: 0.375rem;
    position: relative;
    user-select: none;
    margin-bottom: 0.2rem;
    transition: background 0.15s linear;

    :is(:focus, :hover) {
      background: ${rgba(theme.colors.primary, 0.15)};
      border-radius: 0.375rem;
    }
  `}
`;
