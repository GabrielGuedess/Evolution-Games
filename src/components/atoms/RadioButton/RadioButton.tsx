import { InputHTMLAttributes } from 'react';

import * as S from './RadioButton.styles';

export type RadioButtonProps = {
  title: string;
  icon?: JSX.Element;
} & InputHTMLAttributes<HTMLInputElement>;

export const RadioButton = ({
  title,
  icon,
  id,
  ...props
}: RadioButtonProps) => (
  <S.Wrapper>
    <S.Label htmlFor={id}>
      <S.Radio data-testid={id} id={id} type="radio" {...props} />
      <S.WrapperInfos>
        {icon}
        {!!title && <span>{title}</span>}
      </S.WrapperInfos>
    </S.Label>
  </S.Wrapper>
);
