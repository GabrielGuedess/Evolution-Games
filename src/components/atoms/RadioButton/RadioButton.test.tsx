import { renderWithTheme } from 'utils/tests/helpers';

import { screen } from '@testing-library/react';

import { RadioButton } from './RadioButton';

describe('<RadioButton />', () => {
  it('should render the heading', () => {
    const { container } = renderWithTheme(<RadioButton title="Radio" />);

    expect(screen.getByText(/Radio/i)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
});
