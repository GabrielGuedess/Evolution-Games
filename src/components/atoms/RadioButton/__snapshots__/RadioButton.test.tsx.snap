// Jest Snapshot v1, https://goo.gl/fbAQLP

exports['<RadioButton /> should render the heading 1'] = `
.c1 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  position: relative;
  height: 2rem;
  width: 2rem;
  outline: none;
  margin: 0;
  cursor: pointer;
  border: 0.2rem solid rgba(113,123,138,0.5);
  background: transparent;
  border-radius: 50%;
  display: grid;
  justify-self: end;
  justify-items: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
  -webkit-transition: border 0.5s ease;
  transition: border 0.5s ease;
}

.c1::before,
.c1::after {
  content: '';
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  justify-self: center;
  border-radius: 50%;
}

.c1::before {
  position: absolute;
  width: 100%;
  height: 100%;
  background: #04040C;
  z-index: 1;
  opacity: var(--opacity,1);
}

.c1::after {
  position: relative;
  width: calc(100% / 2);
  height: calc(100% / 2);
  background: #855aff;
  top: var(--y,100%);
  -webkit-transition: top 0.5s cubic-bezier(0.48,1.97,0.5,0.63);
  transition: top 0.5s cubic-bezier(0.48,1.97,0.5,0.63);
}

.c1:checked {
  border: 0.2rem solid #855aff;
  box-shadow: 0 0.5rem 1.5rem rgba(133,90,255,0.25);
}

.c1:checked::after {
  --y: 0%;
  -webkit-animation: edctGW 0.3s ease-out 0.17s;
  animation: edctGW 0.3s ease-out 0.17s;
}

.c1:checked::before {
  --opacity: 0;
}

.c1:checked ~ input[type='radio']::after {
  --y: -100%;
}

.c1:not(:checked)::before {
  --opacity: 1;
  -webkit-transition: opacity 0s linear 0.5s;
  transition: opacity 0s linear 0.5s;
}

.c2 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 1rem;
}

.c2 span {
  text-transform: capitalize;
  font-size: 1.4rem;
}

.c0 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 2rem;
  width: 100%;
  height: 100%;
  padding: 0.375rem;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  margin-bottom: 0.2rem;
  -webkit-transition: background 0.15s linear;
  transition: background 0.15s linear;
}

.c0:is(:focus,:hover) {
  background: rgba(133,90,255,0.15);
  border-radius: 0.375rem;
}

<div
  class=""
>
  <label
    class="c0"
  >
    <input
      class="c1"
      type="radio"
    />
    <div
      class="c2"
    >
      <span>
        Radio
      </span>
    </div>
  </label>
</div>
`;
